File: Blind75-Practice/easy/02-sliding-window/arrays/longest-substring-without-repeating-characters-easy.py
"""
Easy Version: Find First Non-Repeating Character
Difficulty: Easy
Pattern: Hash Map

Given a string s, find the first non-repeating character in it and return its index. If it doesn't exist, return -1.

Example 1:
Input: s = "leetcode"
Output: 0
Explanation: 'l' is the first non-repeating character.

Example 2:
Input: s = "loveleetcode"
Output: 2
Explanation: 'v' is the first non-repeating character.

Example 3:
Input: s = "aabb"
Output: -1
Explanation: No non-repeating character exists.

Time: O(n), Space: O(1) - at most 26 characters
"""


def first_uniq_char(s: str) -> int:
    """
    Hash map approach to count character frequencies
    Time: O(n), Space: O(1)
    """
    char_count = {}

    for ch in s:
        char_count[ch] = char_count.get(ch, 0) + 1

    for i, ch in enumerate(s):
        if char_count[ch] == 1:
            return i

    return -1


# Test cases
if __name__ == "__main__":
    # Test 1
    s1 = "leetcode"
    print(f"First Unique Char: {first_uniq_char(s1)}")  # 0

    # Test 2
    s2 = "loveleetcode"
    print(f"First Unique Char: {first_uniq_char(s2)}")  # 2

    # Test 3
    s3 = "aabb"
    print(f"First Unique Char: {first_uniq_char(s3)}")  # -1

File: Blind75-Practice/easy/02-sliding-window/LongestSubstringWithoutRepeatingCharactersEasy.java
/**
 * Easy Version: Find First Non-Repeating Character
 * Difficulty: Easy
 * Pattern: Hash Map
 *
 * Given a string s, find the first non-repeating character in it and return its index.
 * If it doesn't exist, return -1.
 *
 * Time: O(n), Space: O(1) for 128 ASCII slots
 */
public class LongestSubstringWithoutRepeatingCharactersEasy {

    public static int firstUniqCharAscii(String s) {
        //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

        //

    //

    

        // if (s == null || s.isBlank()) {
        //     return -1;
        // }
        // int[] charFrequency = new int[128];
        // for (int i = 0; i < s.length(); i++) {
        //     char ch = s.charAt(i);
        //     if (ch < 128) {
        //         charFrequency[ch]++;
        //     }
        // }
        // for (int i = 0; i < s.length(); i++) {
        //     char ch = s.charAt(i);
        //     if (ch < 128 && charFrequency[ch] == 1) {
        //         return i;
        //     }
        // }
        // return -1;

        throw new UnsupportedOperationException("Implement firstUniqCharAscii");
    }

    public static void main(String[] args) {
        String s1 = "leetcode";
        System.out.println("First Unique Char: " + firstUniqCharAscii(s1));

        String s2 = "loveleetcode";
        System.out.println("First Unique Char: " + firstUniqCharAscii(s2));

        String s3 = "aabb";
        System.out.println("First Unique Char: " + firstUniqCharAscii(s3));

        String s4 = "a b c a B";
        System.out.println("First Unique (ASCII): " + firstUniqCharAscii(s4));
    }
}
